### PR TITLE
add: showcase start and complete callbacks

### PR DESCRIPTION
**Pull request by vipulasri**
This PR adds functionality to get callbacks on the start and completion of individual showcase view.
Reference: #63

### CODE REVIEW

**Comment by vipulasri** [comment]
@BirjuVachhani Can you have a look at this PR?

**Comment by BirjuVachhani** [comment]
@vipulasri I see that you have upgraded min dart SDK version from `2.1.0` to `2.7.0`. Do you have any specific reason for that?

File: example/lib/main.dart [hunk 1]
```diff
@@ -14,6 +16,12 @@ class MyApp extends StatelessWidget {
       ),
       home: Scaffold(
         body: ShowCaseWidget(
+          onStart: (index, key) => {
```
Comment (BirjuVachhani) [inline example/lib/main.dart:19]: This contains incorrect lambda structure. `=> {}` will create a set. Either use only `=>` or  only `{}`.

File: example/lib/main.dart [hunk 2]
```diff
@@ -14,6 +16,12 @@ class MyApp extends StatelessWidget {
...
+            log('onStart: $index, $key')
+          },
+          onComplete: (index, key) => {
```
Comment (BirjuVachhani) [inline example/lib/main.dart:22]: This contains incorrect lambda structure. `=> {}` will create a set. Either use only `=>` or  only `{}`.

**Comment by vipulasri** [comment]
@BirjuVachhani Yeah, the issue is with the set literals. 

> Set literals weren't supported until version 2.2, but this code is required to be able to run on earlier versions.

File: lib/showcase_widget.dart [hunk 3]
```diff
@@ -29,8 +29,10 @@ import 'package:flutter/material.dart';
 class ShowCaseWidget extends StatefulWidget {
   final Builder builder;
   final VoidCallback onFinish;
+  final Function onStart;
+  final Function onComplete;
```
Comment (BirjuVachhani) [inline lib/showcase_widget.dart:33]: The type `Function` is too generic. Perhaps we should make it statically type safe and restrictive to our use case.

e.g. `Function(int, GlobalKey)` or `typedef ShowCaseEventListener = void Function(int, GlobalKey);`

I prefer the later one. though I am okay with either way.

File: lib/showcase_widget.dart [hunk 4]
```diff
@@ -84,6 +89,18 @@ class ShowCaseWidgetState extends State<ShowCaseWidget> {
     });
   }
 
+  void _onStart() {
+    if (widget.onStart != null && activeWidgetId < ids.length) {
+      widget.onStart(activeWidgetId, ids[activeWidgetId]);
```
Comment (BirjuVachhani) [inline lib/showcase_widget.dart:94]: as these `onStart` and `onComplete` functions are optional, there are chances that these might be null. So handle that case, prefer to use null-safe call to these function invocation.

e.g. `widget.onStart?.call(<args>)`

**Comment by vipulasri** [comment]
@BirjuVachhani I have updated the PR. Thank you for your valuable inputs.

**Comment by BirjuVachhani** [comment]
@vipulasri Thank you for your contribution. Keep up the good work. 👍 
​​​​​​​

**Comment by vipulasri** [comment]
@BirjuVachhani just to confirm, new version of the library will be available today?

**Comment by BirjuVachhani** [comment]
@vipulasri 

I am afraid, I won't be able to publish it today. I'm planning to automate that stuff so it might take 3-4 days. 
Meanwhile you can use it by pointing to this repository.

**Comment by vipulasri** [comment]
Okay, thanks.

**Pull request merged**